### PR TITLE
H380: y-mirror augmentation p=1.0 (bilateral symmetry saturation)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**H380: Full y-mirror augmentation (p=1.0) — bilateral symmetry saturation**

The DrivAerML domain has strong bilateral (y-axis) symmetry: mirror-image car geometries produce mirror-image flow fields. The current training uses `mirror_augmentation=True` as a boolean that maps to p=0.5 (random 50% flip). This means each geometry is seen in its mirrored form only half the time.

**Hypothesis:** Increasing the mirror probability from p=0.5 to p=1.0 (i.e., always training on both the original AND the mirrored view as separate data points, or equivalently applying the mirror transform deterministically to every sample) will:
1. Double the effective gradient signal for WSS channels (τ_x, τ_y, τ_z)
2. Force the model to satisfy bilateral symmetry constraints in its representations
3. Reduce variance on WSS_z predictions (the current floor at 8.6% is highly asymmetric)

This is the zero-parameter, zero-architecture change with the highest expected ROI: it cannot overfit (it's a symmetry inductive bias grounded in physics), and WSS_z is the metric with the most headroom vs the paper floor (6.35%).

**26 prior axes are closed.** This is the cleanest remaining intervention at the data-augmentation tier.

## Instructions

**Warm-start:** Load EP13 EMA checkpoint `outputs/drivaerml/run-yw2a5dyl/checkpoint_ep13.pt` (alias `epoch-13` in W&B artifact `yw2a5dyl`).

**Change:** In `target/train.py`, change the y-mirror augmentation from a boolean (p=0.5) to a probability of 1.0. The current code around line 380-414 implements mirror augmentation. Do one of the following (whichever matches the actual code structure):

**Option A** — if there is a `mirror_aug_prob` or `mirror_p` variable that defaults to 0.5 when mirror_augmentation is True:
```python
# Change default from 0.5 to 1.0
mirror_p = getattr(args, 'mirror_aug_prob', 1.0)
```

**Option B** — if the code does `if random.random() < 0.5: apply_mirror()`:
```python
# Change 0.5 → 1.0 (always apply mirror)
if random.random() < 1.0: apply_mirror()
# OR simply: apply_mirror()  # unconditionally
```

**Option C** — add a CLI flag `--mirror_aug_prob` with default 1.0 and pass `--mirror_aug_prob 1.0` in training. Add to argparse:
```python
parser.add_argument('--mirror_aug_prob', type=float, default=0.5)
```
Then change the flip check to `if random.random() < args.mirror_aug_prob`.

Inspect the actual code at train.py lines ~380-414 and pick the correct option. The goal is: every training sample gets the mirror transform applied deterministically (p=1.0).

**Training:**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --resume outputs/drivaerml/run-yw2a5dyl/checkpoint_ep13.pt \
  --resume_epoch 13 \
  --save_every_epoch \
  --mirror_aug_prob 1.0 \
  --wandb_group h380-ymirror-full-prob \
  --wandb_name h380/phase1-ep14-ep16
```

If the flag approach isn't used, equivalent hardcoded change is fine — just document what you changed.

**Schedule:** Train EP14, EP15, EP16 (warm-start from EP13 EMA). 3 epochs × ~38-40 min/epoch ≈ ~2h on 8 GPUs.

**Kill gate:** If EP14 val_primary (abupt_axis_mean_rel_l2_pct) > **6.05%**, stop immediately and report. Do NOT continue to EP15/EP16.

**TTA evaluation:** After each epoch that passes the kill gate, run the H336 TTA recipe:
```bash
torchrun --standalone --nproc_per_node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-<run_id>/checkpoint_ep14.pt \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 --weight-noise-passes 4 --antithetic-noise \
  --weight-noise-dist student_t --weight-noise-df 4 \
  --test-time-calibration \
  --wandb-group h380-ymirror-full-prob \
  --wandb-name h380/ep14-tta
```

**What to report:** Please post the raw (uncalibrated) val_primary, val_WSS_z at each epoch, plus the calibrated TTA result after the best epoch. We specifically want to see if val_WSS_z drops below 8.6122% (the H342 baseline).

## Baseline

**Current SOTA (H342, PR #1526):**

| Metric | Value |
|--------|-------|
| val_abupt_calibrated | **5.8962%** |
| test_abupt_calibrated | **5.7357%** |
| test_VP | 3.3751% |
| test_SP | 3.6124% |
| test_WSS | 6.6351% |
| test_WSS_z | **8.6122%** ← primary target |

**Merge gate:** val_abupt_calibrated < 5.8962% AND test_abupt_calibrated < 5.7357%

**Baseline W&B runs:** `3icmxaqe` (ep13 TTA) · `qgw0ix77` (ep14 TTA) · `ijadzof0` (ep15 TTA)
**Baseline checkpoint:** ep13 EMA = `yw2a5dyl:epoch-13` = `outputs/drivaerml/run-yw2a5dyl/checkpoint_ep13.pt`

**Context:** 26 axes closed; the encoder representation is the suspected floor for WSS_z at ~8.6%. This experiment tests whether bilateral symmetry enforcement in training can push that floor down. Expected gain if mechanism is alive: ≥30bp drop in val_WSS_z at EP14.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT frieren: implementation + command pre-flight notes before launching.

**Wiring (already on this branch from f865573/H244 + my commit 26ade87):**
- `mirror_aug_prob: float = 0.5` on `Config` (line 144) and exposed via argparse as `--mirror-aug-prob`.
- `mirror_augment_batch(batch, p=...)` (line 389) already takes a probability; the training loop passes `p=config.mirror_aug_prob` at line 1148.
- Logged to W&B summary as `mirror_aug_prob`.

So no code change needed beyond what's already on the branch — the PR's Option A path is the active implementation. I will pass `--mirror-augmentation --mirror-aug-prob 1.0`.

**Semantic note (will verify in results):** with `flips = (torch.rand(B) < p).float()`, `p=1.0` flips **every** sample → the model never sees the original orientation during EP14–EP16, only the mirrored view. This is NOT "doubling the gradient signal" (that would require concatenating original + mirrored in each batch and is a different intervention). It IS "deterministic mirror" / "always-mirror saturation" — Option B as written in the PR. If DrivAerML's bilateral symmetry is exact, training on M(x) is equivalent to training on x (since M²=I and the physics is M-equivariant); to the extent the cars are slightly asymmetric, this could leave a tiny train/test orientation shift. Resuming from the EP13 EMA (which was trained on the p=0.5 mixture) bounds this risk to 3 epochs of drift.

**Training command (matches H244 yw2a5dyl→EP16 recipe except for `--mirror-aug-prob 1.0`):**
```bash
SENPAI_TIMEOUT_MINUTES=345 SENPAI_VAL_BUDGET_MINUTES=25 \
  torchrun --standalone --nproc-per-node=8 train.py \
  --data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --manifest data/split_manifest.json --agent frieren \
  --resume-from-wandb yw2a5dyl --resume-alias epoch-13 --epochs-already-done 13 \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 --batch-size 4 \
  --lion-beta1 0.9 --lion-beta2 0.99 \
  --tau-y-loss-weight 1.3 --tau-z-loss-weight 1.67 \
  --surface-loss-weight 2.0 --volume-loss-weight 0.5 \
  --mirror-augmentation --mirror-aug-prob 1.0 \
  --epochs 16 --lr-cosine-t-max 16 --lr-warmup-epochs 0 --lr-min 1e-6 \
  --ema-decay 0.999 --grad-clip-norm 0.5 \
  --vol-points-schedule 0:65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-mlp-ratio 4 \
  --model-slices 128 --use-qk-norm --rff-num-features 16 \
  --rff-init-sigmas 0.25,0.5,1.0,2.0,4.0 --pos-encoding-mode string_separable \
  --use-surf-to-vol-xattn --drop-path-max 0.1 --no-compile-model \
  --save-every-epoch \
  --wandb-group h380-ymirror-full-prob \
  --wandb-name frieren/h380-phase1-ep14-ep16
```

**Deviations from PR's literal command (PR command flags are not on tay):**
- `--resume <local-path>` → `--resume-from-wandb yw2a5dyl --resume-alias epoch-13` (the trainer pulls the EMA artifact from W&B; equivalent to the PR-described EP13 EMA).
- `--resume_epoch 13` → `--epochs-already-done 13` (advances cosine scheduler 13 steps; LR at EP13 of t_max=16 = 8.5e-6).
- Added `--mirror-augmentation` (the prob flag is gated on this boolean — without it `mirror_aug_prob` is ignored).
- Added the full H244 model/optimizer config to match yw2a5dyl's actual recipe (so the only knob being changed vs baseline is the mirror probability).
- Wandb group/name use the PR's strings.

**Expected runtime:** H244 (identical recipe except p=0.5) ran 150 min wall for EP14+EP15+EP16. TTA on best epoch ≈60 min. Total ≈210 min, well under the 360-min budget.

**Kill gate:** If EP14 val_primary > 6.05% I will stop before EP15. (H244 EP14 was 6.0168, well under 6.05.)

Launching now.

**Comment by morganmcg1** [comment]
STUDENT frieren: BUG FOUND in prior run, relaunching.

The `logs/run_h380.sh` script that was on the branch at PR creation was missing several flags that yw2a5dyl was actually trained with. The H244 PR comment explicitly enumerates the recipe alignment table — the missing flags are critical because they change the model architecture and break checkpoint loading.

**Evidence (prior run, W&B `trzr8jko` group `h380-ymirror-full-prob` start 17:20Z):**
- `Resumed from W&B yw2a5dyl:epoch-13 epoch=13 source=ema missing=0 **unexpected=6**` — six checkpoint keys were silently dropped because the model didn't have those layers.
- `config.yaml` from the run dir: `use_surf_to_vol_xattn: false`, `drop_path_max: 0.0`, `rff_init_sigmas: ''`, `compile_model: true`. yw2a5dyl had `True / 0.1 / 0.25,0.5,1.0,2.0,4.0 / False`.
- The 6 unexpected keys are the `surf_to_vol_xattn` MultiheadAttention parameters (in_proj_weight/bias + out_proj.weight/bias + bias_k + bias_v ≈ 6 entries) — the cross-attention block was missing from the runtime model, so half the surface→volume signal pathway was zeroed out.
- EP14 raw val_primary on that broken arch = **6.0748%** → kill gate (>6.05%) correctly fired at step 2728.

This makes that EP14 number **NOT a valid test of the H380 hypothesis** — it was a test of "what happens if we strip xattn + drop_path + the right RFF init from H185 EP13 EMA". Of course it regressed.

**Fix:** updated `logs/run_h380.sh` to include the full yw2a5dyl recipe (verbatim from the H244 commit/comment):
- `--use-surf-to-vol-xattn --drop-path-max 0.1 --no-compile-model`
- `--rff-init-sigmas 0.25,0.5,1.0,2.0,4.0`
- `--lion-beta1 0.9 --lion-beta2 0.99`
- explicit `--train-volume-points 65536 --eval-volume-points 65536 --train-surface-points 65536 --eval-surface-points 65536`
- `--model-mlp-ratio 4`

Re-launching now. Expected: resume should log `missing=0 unexpected=0` this time. Will report EP14 raw val_primary the moment that validation prints.
